# Evaluation of dada2 pipeline on all sequencing labs

```{r package_options, include=FALSE}
knitr::opts_knit$set(progress = TRUE, verbose = TRUE)
# There are lots of messages about the phylo object being defined by multiple
# packages, which we do not need to see
knitr::opts_chunk$set(message=FALSE)
```

```{r setup, include=FALSE}
library(stringr)
library(phyloseq); library(dada2);packageVersion("dada2")
library(Biostrings); library(ShortRead)
library(ggplot2)

## Metadata for each sequencing lab
df = readRDS(file.path('~/active_research/metagenomics_calibration/mbqc/data',
             'blinded_sequence_table.rds'))
# df$sequencing_wetlab = sapply(df$sequencing_wetlab, str_sub, -1)
path = "~/data/mbqc/blinded_sequence_data"
setwd(path)
            
## Primer sequences
F515 <- "GTGCCAGCMGCCGCGGTAA"
R806 <- "GGACTACHVGGGTWTCTAAT"
F318 <- "ACTCCTACGGGAGGCAGCAG"
F515_HLC <- "GTGTGCCAGCMGCCGCGGTAA"
# HL-C used an F515 primer with an extra 2bp inserted in after the first 2bp of
# the F515 primer sequence quoted in the paper (see
# mbqc_handling_protocols.xlsx)

## Filenames of raw data, used for quality profiles
# All labs have paired end reads except for HL-I
fns.all = list.files(path, recursive=TRUE)
fnIs = grep(pattern="raw/HL-I/.*_R1.fastq.gz$", fns.all, value=TRUE)
fnFs = grep(pattern="raw/HL-[^I]/.*_R1.fastq.gz$", fns.all, value=TRUE)
fnRs = grep(pattern="raw/HL-[^I]/.*_R2.fastq.gz$", fns.all, value=TRUE)
names(fnIs) <- sapply(strsplit(fnIs, "[_/]"), "[", 3)
names(fnFs) <- sapply(strsplit(fnFs, "[_/]"), "[", 3)
names(fnRs) <- sapply(strsplit(fnRs, "[_/]"), "[", 3)
identical(names(fnRs), names(fnFs))

# List of labs and samples for each lab
split.samples = split(df$Bioinformatics.ID, df$sequencing_wetlab)
names(split.samples) = str_sub(names(split.samples), -1)
seqlabs = names(split.samples)
num.samples = sapply(split.samples,length)
# Load up to three test samples for each lab to check quality # profiles.
set.seed(1)
test.samples = mapply(sample, split.samples, sapply(num.samples, min, 3))

# Load chosen trim.params
trim.params = readRDS(file.path(path, 'dada_out', paste0('trim_params', '.rds')))

## Load quality profiles of test samples
qpFs = readRDS(file.path(path, 'qc', paste0('quality_profiles_Fs', '.rds')))
qpRs = readRDS(file.path(path, 'qc', paste0('quality_profiles_Rs', '.rds')))
qpIs = readRDS(file.path(path, 'qc', paste0('quality_profiles_labI', '.rds')))
## Load read lengths of test samples
rlFs = readRDS(file.path(path, 'qc', paste0('read_lengths_Fs', '.rds')))
rlRs = readRDS(file.path(path, 'qc', paste0('read_lengths_Rs', '.rds')))
rlIs = readRDS(file.path(path, 'qc', paste0('read_lengths_labI', '.rds')))

# Compute min overlap assuming max length 255
overlaps = with(trim.params, truncF + truncR - trimF - trimR - 255)
names(overlaps) = rownames(trim.params)

# Compute tracking table
track.li = lapply(file.path(path, 'dada_out', 
                            paste0('track_', seqlabs, '.rds')),
                  function(x) as.data.frame(readRDS(x)))
names(track.li) = seqlabs
track.li$I$merged = NA
track.df = do.call(rbind, unname(track.li))
# track.df$HL = unlist(df[rownames(track.df), 'sequencing_wetlab'])
# Note: filtered==denoised and tabled==merged, at least for this data.
track.df = within(track.df, {
    f.nonchim = nonchim/tabled
    f.merged = merged/denoised
    f.filtered = filtered/input
    f.total = nonchim/input
    })
track.bylab = aggregate(track.df, df[rownames(track.df), 'sequencing_wetlab'],
                        mean)
rownames(track.bylab) = sapply(track.bylab$sequencing_wetlab, str_sub, -1)
track.bylab = subset(track.bylab, select=-sequencing_wetlab)
track.bylab$num.samples = sapply(split.samples, length)

## Info for each lab
hl.df = subset(df, select=c(sequencing_wetlab, seq_machine, paired_end_reads,
                            read_length, frac_quality_bases, phiX_frac))
hl.df = unique(hl.df)
rownames(hl.df) = sapply(hl.df$sequencing_wetlab, str_sub, -1)
hl.df = subset(hl.df, select=-sequencing_wetlab)
# Fix wrong data
hl.df['F', 'paired_end_reads'] = 'Yes'
hl.df['I', 'paired_end_reads'] = 'No'
#
med.rlF = lapply(rlFs, function (x) median(unlist(x)))
med.rlR = lapply(rlRs, function (x) median(unlist(x)))
med.rlF$I = median(unlist(rlIs))
med.rlR$I = NA
hl.df$read_length.F = med.rlF[rownames(hl.df)]
hl.df$read_length.R = med.rlR[rownames(hl.df)]
```

Basic info for each lab. Read lengths are the median in three test files, and
differ from what was reported for labs F and I. Similarly, the pair-end status
was incorrectly reported for labs F and I. The fields frac_quality_bases and
phiX_frac are what was reported to MBQC.
```{r, echo=FALSE, R.options=list(width=100)}
print(subset(hl.df, select=c(seq_machine, paired_end_reads, 
                             read_length.F, read_length.R,
                             frac_quality_bases, phiX_frac)))
```

Trim and truncation parameters used, with min overlap between the forward and
reverse reads assuming a max amplicon length of 255 bp:
```{r, echo=FALSE}
trim.params$overlap = overlaps
print(trim.params)
```

Filtering parameters used: `maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE`
except for lab M, where I raised the error rate to `maxEE=c(10,10)`.

Summary of the runs on all labs, averaging across samples in the lab. The
number of samples, average number of input reads per sample, and the average
fraction kept at the filtering, merging, and chimera-removal steps, and average
fraction kept overall.
```{r, echo=FALSE}
print(track.bylab[, c('num.samples', 'input', 'f.filtered', 'f.merged',
                      'f.nonchim', 'f.total')],
      digits=2)
```


## Quality profiles

```{r, eval=FALSE, echo=FALSE, results='asis'}
for (lab in seqlabs) {
    cat('  \n###Sequencing lab', lab)
    cat('  \n')
    tps = unlist(trim.params[lab, ])
    if (! lab %in% c('C', 'I')) {
        print(qpFs[[lab]] + geom_vline(xintercept=tps['truncF']))
        print(qpRs[[lab]] + geom_vline(xintercept=tps['truncR']))
    }
    if (lab == 'C') {
        print(qpFs[[lab]] + geom_vline(xintercept=tps['trimF']) + 
            geom_vline(xintercept=tps['truncF']))
        print(qpRs[[lab]] + geom_vline(xintercept=tps['trimR']) + 
            geom_vline(xintercept=tps['truncR']))
    }
    if (lab == 'I') {
        print(qpIs + geom_vline(xintercept=tps['trimF']) + 
            geom_vline(xintercept=tps['truncF']))
    }
    cat('  \n')
}
```

<!-- Old -->
```{r, echo=FALSE, eval=FALSE}
for (lab in seqlabs[-c(3,7)]) {
    print(paste('Sequencing lab', lab))
    tps = unlist(trim.params[lab, ])
    print(paste('Trim parameters: trimLeft=', tps[1], tps[2], 
                ', truncLen=', tps[3], tps[4]))
    print('Read lengths (Forward)')
    print(summary(c(unlist(rlFs[lab]))))
    print('Read lengths (Reverse)')
    print(summary(c(unlist(rlRs[lab]))))
    print('Quality Profiles')
    if (lab != 'C') {
        print(qpFs[[lab]] + geom_vline(xintercept=tps['truncF']))
        print(qpRs[[lab]] + geom_vline(xintercept=tps['truncR']))
    }
    if (lab == 'C') {
        print(qpFs[[lab]] + geom_vline(xintercept=tps['trimF']) + 
            geom_vline(xintercept=tps['truncF']))
        print(qpRs[[lab]] + geom_vline(xintercept=tps['trimR']) + 
            geom_vline(xintercept=tps['truncR']))
    }
    # Results from pipeline
    print('Read tracking')
    track = readRDS(file.path(path, 'dada_out', paste0('track_', lab, '.rds')))
    print(head(track, 6))
}
```

