---
title: "Import Metaphlan2 analysis of Costea2017 Phase 3 data into phyloseq"
output:
  html_document:
    toc: false
    toc_float: false
    self_contained: true
    code_folding: show
---

# Setup

Starting point: Files produced by metaphlan2 should be in the folder
`$DATA_DIR/costea2017/intermediate`.

Load packages.
```{r libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(phyloseq)
library(metaphlanr)
```
Set paths.
```{r path}
dotenv::load_dot_env("../.env")
script_path <- getwd()
data_path <- file.path(Sys.getenv("DATA_DIR"),
    "costea2017")
tree_path <- file.path(Sys.getenv("DATA_DIR"),
    "taxonomy")
```

## Prepare the sample metadata

Load the sample metadata generated by `sample_metadata.R`.
```{r}
sam <- read_csv(file.path(script_path, "sample_metadata.csv"))
```

We'll simplify the sample names and remove uncessary metadata fields.
```{r}
# Remove unneccessary metadata fields
sam0 <- sam %>%
    select(Sample, Protocol, Individual, SRA_run)
# Remove the two extra specimens only extracted by protocol Q
sam0 <- sam0 %>%
    filter(!(Individual %in% c("A", "B")))
# Shorten `Individual` label of "Mock-only" to just "M"
sam0 <- sam0 %>%
    mutate(Individual = str_sub(Individual, 1, 1))
# Create short, meaningful sample names
sam0 <- sam0 %>%
    mutate(Sample = paste0(Protocol, Individual))
```

Phyloseq wants a plain dataframe with row names set to the sample names. 
```{r}
# Create a data frame for phyloseq
sam1 <- sam0 %>%
    select(-Sample, -SRA_run) %>%
    as.data.frame()
row.names(sam1) <- sam0$Sample
sam.ps <- sample_data(sam1)
```

## Prepare phylogeny

Use the tree included in the Waldron lab's Bioconductor package
`curatedMetagenomicData`.

<!-- https://groups.google.com/forum/#!msg/metaphlan-users/qEIqLrI37g4/uwFtVeFmAgAJ -->

```{r}
# Check if tree file already exists; if not, download it to data_path/taxonomy
tree_url <-
    "https://github.com/waldronlab/curatedMetagenomicData/blob/master/inst/extdata/metaphlan2_selected.tree.reroot.nwk.bz2?raw=true"
if (!file.exists(file.path(tree_path,
            "metaphlan2_selected.tree.reroot.nwk.bz2"))) {
    download.file(tree_url, 
        destfile = file.path(tree_path,
            "metaphlan2_selected.tree.reroot.nwk.bz2"),
        method = "auto")
}
tree <- ape::read.tree(file.path(tree_path,
        "metaphlan2_selected.tree.reroot.nwk.bz2"))
tree
```

Check the format of the tip labels. All are clade strings at the species level,
ending in a GCF identifier, except for a few with no taxonomy info.
```{r}
head(tree$tip.label)
tree$tip.label %>%
    {.[!str_detect(., "s__")]}
```

Simplify the tip names to just the species. Need to first remove the few
with no taxonomy.
```{r}
# Prune tips with no taxonomy
tree0 <- str_detect(tree$tip.label, "s__") %>%
    {tree$tip.label[!.]} %>%
    ape::drop.tip(tree, .)
# Simplify tip names
tree_species <- tree0$tip.label %>% 
    str_extract("(?<=s__)\\w+(?=\\|)")
sum(is.na(tree_species))
tree0$tip.label <- tree_species
tree0
```

# Relative abundance profiles (Metaphlan2 default analysis)

## Prepare species-level OTU tables from metaphlan output

Starting point is the file `merged_abundance_table.tsv` produced by the
metaphlan helper script.

We need to get just the species entries and simplify to just the species names,
while saving the taxonomy and parsing into a taxonomy table.
```{r}
st <- readr::read_tsv(file.path(data_path, "intermediate", "rel_ab",
        "merged_abundance_table.tsv"))
st <- st %>%
    rename(Clade = ID)
# Filter to species and get species name
st0 <- st %>%
    filter(str_detect(Clade, "s__"), !str_detect(Clade, "t__")) %>%
    mutate(Species = str_extract(Clade, "(?<=s__)\\w+$")) %>%
    select(Species, everything())
# Check that relative abundances add (roughly) to 100
st0 %>% 
    summarize_at(vars(-Species, -Clade), sum) %>%
    gather(key = "Sample", value = "Sum") %>%
    summary()
# Get into taxa-are-columns format
st1 <- st0 %>%
    select(-Clade) %>%
    gather(key = "Sample", value = "Abundance", -Species) %>%
    spread(key = Species, value = "Abundance")
# Simplify sample names
st1 <- st1 %>%
    mutate(Sample = str_extract(Sample, "ERR[0-9]+"))
```
Use sample names matching the sample data.
```{r}
species <- select(st1, -Sample) %>% colnames
anyDuplicated(species) == 0
st2 <- left_join(sam0, st1, by = c("SRA_run" = "Sample")) %>%
    select(Sample, species)
```
Make a phyloseq otu_table object.
```{r}
# Get a sequence matrix with samples as rows, species as columns
mat <- select(st2, -Sample) %>%
    as("matrix")
rownames(mat) <- st2$Sample
st.ps <- otu_table(mat, taxa_are_rows = FALSE)
```

## Prepare taxonomy

```{r}
# Get the tax table
tax <- metaphlanr::parse_taxonomy(st0$Clade) %>%
    select(-Strain)
# Taxonomy for phyloseq
taxmat <- tax %>%
    select(Kingdom:Species) %>%
    as("matrix")
rownames(taxmat) <- taxmat[,"Species"]
tax.ps <- tax_table(taxmat)
```

## Make phyloseq object

```{r}
ps <- phyloseq(sam.ps, st.ps, tax.ps)
sample_sums(ps)
```

## Add phylogeny


First, let's compare the species names in the tree and the metaphlan output.
```{r}
taxa <- taxa_names(ps)
setdiff(taxa, tree_species)
taxa %>% str_subset("Streptococcus_mitis")
tree_species %>% str_subset("Streptococcus_mitis")
taxa %>% str_subset("Lactobacillus_casei")
tree_species %>% str_subset("Lactobacillus_casei")
```
The tree only contains bacteria and archaea, while the metaphlan output
includes viruses; the metaphlan output includes taxa labeled as
"Genus_unclassified", corresponding to abundance assigned to a genus beyond the
species of that genus, which of course aren't on the tree; and some of these
species in the metaphlan output have simpler names than in the tree. Since
there are just two of these in our data, I will manually fix the names.
```{r}
tree_species0 <- tree_species %>% 
    replace(., . == "Streptococcus_mitis",
        "Streptococcus_mitis_oralis_pneumoniae") %>%
    replace(., . == "Lactobacillus_casei", "Lactobacillus_casei_paracasei")
setdiff(taxa_names(ps), tree_species0)
tree0$tip.label <- tree_species0
# Merge the tree into the phyloseq object
ps.tree <- merge_phyloseq(ps, tree0)
ps
ps.tree
```
Viruses and the "unclassified" taxa are removed in the `ps.tree` object.

## Save phyloseq objects

Save both phyloseq objects; the version without the tree is needed for looking
at viruses and the "unclassified"'s.
```{r}
saveRDS(ps, file.path(data_path, "final", 
        paste0(Sys.Date(), "_phyloseq_phase3_ra.Rds")))
saveRDS(ps.tree, file.path(data_path, "final", 
        paste0(Sys.Date(), "_phyloseq_phase3_ra_tree.Rds")))
```

