---
title: "Import Brooks2015 species table into phyloseq"
output:
  html_document:
    toc: false
    toc_float: false
    self_contained: true
    code_folding: show
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_knit$set(progress=TRUE, verbose=TRUE)
# Global chunk options
knitr::opts_chunk$set(cache=FALSE, echo=TRUE, warning=FALSE,
    message=FALSE, include=TRUE)
# R options
# options(width=100)
options(stringsAsFactors = FALSE)
```

Here we will download the supplementary files and create a phyloseq object with
the OTU table created by the authors.

# Setup

Load packages.
```{r libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(phyloseq)
```
Set paths.
```{r path}
dotenv::load_dot_env("../.env")
script_path <- getwd()
data_path <- file.path(Sys.getenv("DATA_DIR"),
    "brooks2015")
tax_path <- file.path(Sys.getenv("DATA_DIR"), "taxonomy", "gtdb")
```

```{r}
prepend_date <- function (fn) {
    paste(Sys.Date(), fn, sep="_")
}
```

# Download supplemental files

https://bmcmicrobiol.biomedcentral.com/articles/10.1186/s12866-015-0351-6

"Additional file 2 Experimental design. Table of the prescribed mixing
proportions, plate, and barcode for the experiments mixing equal proportions of
cells, DNA, and PCR product."
<!-- https://static-content.springer.com/esm/art%3A10.1186%2Fs12866-015-0351-6/MediaObjects/12866_2015_351_MOESM2_ESM.csv -->

"Additional file 10 Table of above-threshold counts. Above-threshold counts for
each sample in the experiments mixing equal amounts of cells, DNA, and PCR
product."
<!-- https://static-content.springer.com/esm/art%3A10.1186%2Fs12866-015-0351-6/MediaObjects/12866_2015_351_MOESM10_ESM.csv -->

"Additional file 11 Table of below-threshold counts. Below-threshold counts for
each sample in the experiments mixing equal amounts of cells, DNA, and PCR
product."
<!-- https://static-content.springer.com/esm/art%3A10.1186%2Fs12866-015-0351-6/MediaObjects/12866_2015_351_MOESM11_ESM.csv -->


```{r}
file_num <- c(2, 10, 11)
urls <- paste0("https://static-content.springer.com/esm/",
    "art%3A10.1186%2Fs12866-015-0351-6/MediaObjects/",
    "12866_2015_351_MOESM", file_num, "_ESM.csv")
fns <- file.path(data_path, "supplementary_files", 
    paste0("AdditionalFile", file_num, ".csv"))
if (!dir.exists(file.path(data_path, "supplementary_files"))) {
    dir.create(file.path(data_path, "supplementary_files"), recursive = TRUE)
    walk2(urls, fns, download.file)
}
```

# Build the species/OTU table

This code is copied from pdf rmarkdown output provided by the authors as
[Additional file 7](https://static-content.springer.com/esm/art%3A10.1186%2Fs12866-015-0351-6/MediaObjects/12866_2015_351_MOESM7_ESM.pdf)

```{r brooks}
# Define column names.
organismsDesign <- c("Gvaginalis", "Avaginae", "Lcrispatus", "Liners",
    "Pbivia", "Samnii", "GroupBStrep")
organismsResults <- c("Gardnerella.vaginalis", "Atopobium.vaginae",
    "Lactobacillus.crispatus_cluster", "Lactobacillus.iners",
    "Prevotella.bivia", "Sneathia.amnii", "Streptococcus.agalactiae")
organismsForPlots <- c("G. vaginalis", "A. vaginae", "L. crispatus", 
    "L. iners", "P. bivia", "S. amnii", "S. agalactiae")
# Genome size and copy numbers obtained from NCBI.
genomeSize <- c(1.65, 1.43, 2.04, 1.3, 2.47, 1.34, 2.2)
copyNumbers <- c(2,1,4,1,1,1,7)
# Read above-threshold counts data. Remove last (blank) column.
countdata <- read.table(
    file.path(data_path, "supplementary_files", "AdditionalFile10.csv"), 
    sep=",", header=TRUE, row.names=1, 
    colClasses=c("character", rep("numeric", 46)))
countdata <- countdata[,-c(ncol(countdata))]
atOrganisms <- colnames(countdata)
# Summarize the number of counts per sample.
atcounts <- apply(countdata,1,sum)
summary(atcounts)
# Calculate the total number above-threshold reads.
sum(atcounts)
# Read the below-threshold counts data.
btdata <- read.table(
    file.path(data_path, "supplementary_files", "AdditionalFile11.csv"), 
    sep=",", header=TRUE, row.names=1)
btdata <- btdata[,-c(ncol(btdata))]
btOrganisms <- sub("BT","", colnames(btdata))
# Parse the sample IDs to get the plate and barcode numbers.
m <- regexec("([1-6])_([0-9]+)" , rownames(countdata))
matchlist <- regmatches(rownames(countdata), m)
matchlistmatrix <- matrix(unlist(matchlist), ncol=3, byrow=TRUE)
Plate <- as.numeric(matchlistmatrix[,2])
Barcode <- as.numeric(matchlistmatrix[,3])
# Join above-threshold counts with plate and barcode numbers.
atdata <- data.frame(countdata, Plate, Barcode)
atdata <- atdata[order(Plate,Barcode),]
# Join below-threshold counts with plate and barcode numbers.
btdata <- data.frame(btdata, Plate, Barcode)
btdata <- btdata[order(Plate, Barcode),]
# Read in the design from AdditionalFile2.csv and merge with the
# above-threshold counts data.
design <- read.table(
    file.path(data_path, "supplementary_files", "AdditionalFile2.csv"), 
    sep=",", header=TRUE, row.names=1,
    colClasses = c(rep("character",2), rep("numeric",9)) )
alldata <- merge(design, atdata, by=c("Plate", "Barcode"), all=TRUE)
alldata[,4:ncol(alldata)] <- sapply(alldata[,4:ncol(alldata)], as.numeric)
# Label each sample according to the experiment. Experiment 1 mixed equal
# numbers of cells, Experiment 2 mixed equal DNA, Experiment 3 mixed equal PCR
# product.
experiment <- numeric(nrow(alldata))
experiment[(alldata$Plate == 1) | (alldata$Plate == 2)] <- 1
experiment[(alldata$Plate == 3) | (alldata$Plate == 4)] <- 2
experiment[(alldata$Plate == 5) | (alldata$Plate == 6)] <- 3
experiment <- factor(experiment)
# Get the number of above-threshold reads classified as belonging to taxa that
# were not in the study.
otherData <- alldata[,-match(c("Plate","Barcode","Experiment", organismsDesign,
        organismsResults), names(alldata))]
otherCounts <- apply(as.matrix(otherData),1,sum)
sum(otherCounts)
```

```{r}
# Get the number of below-threshold reads classified as belonging to taxa not
# in the study.
allbtdata <- merge(btdata, atdata, by=c("Plate", "Barcode"), all=TRUE)
allbtdata <- sapply(allbtdata, as.numeric)
btexp1 <- allbtdata[(allbtdata[,"Plate"] == 1) | (allbtdata[,"Plate"] == 2),]
btexp2 <- allbtdata[(allbtdata[,"Plate"] == 3) | (allbtdata[,"Plate"] == 4),]
btexp3 <- allbtdata[(allbtdata[,"Plate"] == 5) | (allbtdata[,"Plate"] == 6),]
totalcounts1 <- apply(btexp1,1,sum)
totalcounts2 <- apply(btexp2,1,sum)
totalcounts3 <- apply(btexp3,1,sum)
btCounts1 <- btexp1[,-match(names(atdata), colnames(btexp1))]
btCounts2 <- btexp2[,-match(names(atdata), colnames(btexp2))]
btCounts3 <- btexp3[,-match(names(atdata), colnames(btexp3))]
btResultsOrganisms <- paste(organismsResults, "BT", sep="")
btNotResultsCounts1 <- sum(btCounts1[,-match(btResultsOrganisms,
        colnames(btCounts1))])
btNotResultsCounts2 <- sum(btCounts2[,-match(btResultsOrganisms,
        colnames(btCounts2))])
btNotResultsCounts3 <- sum(btCounts3[,-match(btResultsOrganisms,
        colnames(btCounts3))])
btNotResultsCounts1 + btNotResultsCounts2 + btNotResultsCounts3
# Get total number of reads (above- and below-threshold).
sum(totalcounts1)+sum(totalcounts2)+sum(totalcounts3)
# Normalize the above-threshold data to proportions.
dataNorm <- data.frame(alldata[,c(organismsDesign, organismsResults)],
    otherCounts)
Normcounts <- apply(dataNorm[,c(organismsResults, "otherCounts")],1,sum)
classcounts <- Normcounts
dataNorm[,c(organismsResults, "otherCounts")] <- dataNorm[,c(organismsResults,
    "otherCounts")]/Normcounts
# Adjust counts data for copy number and genome size.
dataNorm[experiment == 1,organismsResults] <- t(t(dataNorm[experiment ==
        1,organismsResults])/copyNumbers)
dataNorm[experiment == 2,organismsResults] <- t(t(dataNorm[experiment ==
        2,organismsResults])*(genomeSize/copyNumbers))
# Re-normalize to proportions.
Normcounts <- apply(dataNorm[,c(organismsResults, "otherCounts")],1,sum)
dataNorm[,c(organismsResults, "otherCounts")] <- dataNorm[,c(organismsResults,
    "otherCounts")]/Normcounts
# Summarize the proportion of above-threshold reads classified as belonging to
# taxa not in the study for each sample.
summary(dataNorm$otherCounts)
```

```{r}
# Make a data frame for each experiment.
classcounts1 <- classcounts[experiment == 1]
classcounts2 <- classcounts[experiment == 2]
classcounts3 <- classcounts[experiment == 3]
exp1Norm <- dataNorm[experiment == 1,]
exp2Norm <- dataNorm[experiment == 2,]
exp3Norm <- dataNorm[experiment == 3,]
# Re-order the rows according to the design.
classcounts1 <- classcounts1[with(exp1Norm, order(Gvaginalis, Avaginae,
        Lcrispatus, Liners, Pbivia, Samnii, GroupBStrep))]
classcounts2 <- classcounts2[with(exp2Norm, order(Gvaginalis, Avaginae,
        Lcrispatus, Liners, Pbivia, Samnii, GroupBStrep))]
classcounts3 <- classcounts3[with(exp3Norm, order(Gvaginalis, Avaginae,
        Lcrispatus, Liners, Pbivia, Samnii, GroupBStrep))]
exp1Norm <- exp1Norm[with(exp1Norm, order(Gvaginalis, Avaginae, Lcrispatus,
        Liners, Pbivia, Samnii, GroupBStrep)),]
exp2Norm <- exp2Norm[with(exp2Norm, order(Gvaginalis, Avaginae, Lcrispatus,
        Liners, Pbivia, Samnii, GroupBStrep)),]
exp3Norm <- exp3Norm[with(exp3Norm, order(Gvaginalis, Avaginae, Lcrispatus,
        Liners, Pbivia, Samnii, GroupBStrep)),]
```

# Create a phyloseq object

Now we want to create a `phyloseq` to use for our analysis. We want an otu
table unadjusted by 16S copy number or genome size, since we want to include
all sources of bias and be able to control such adjustments later.
```{r}
# Delete variables to conflicting with the tibble column names
remove(Plate, Barcode)
tb <- alldata %>% as_tibble %>%
    mutate(Plate = as.integer(Plate), Barcode = as.integer(Barcode)) %>%
    arrange(Plate, Barcode)
tb <- tb %>%
    add_column(Sample = paste0('s', tb$Plate, '-', tb$Barcode), 
        .before = 1)
# Check that the plates match the experiments as expected
tb %>% select(Plate, Experiment) %>% table
# Sample metadata
sam <- tb %>%
    select(Sample, Plate, Barcode, Experiment, organismsDesign)
num_species <- sam[, organismsDesign] %>% apply(1, function (x) sum(x>0))
sam <- sam %>%
    add_column(Num_species = num_species, .after = "Experiment")
# Sequence table
st <- tb %>%
    select(Sample, organismsResults)
# Use consistent taxa names
# taxa <- c("G.vaginalis", "A.vaginae", "L.crispatus", "L.iners", "P.bivia",
#     "S.amnii", "S.agalactiae")
taxa <- c("Gardnerella_vaginalis", "Atopobium_vaginae",
    "Lactobacillus_crispatus", "Lactobacillus_iners",
    "Prevotella_bivia", "Sneathia_amnii", "Streptococcus_agalactiae")
names(sam) <- names(sam) %>% {c(.[1:5], taxa)}
names(st) <- names(st) %>% {c(.[1], taxa)}
# Build a phyloseq object
sam <- sam %>%
    select(-Sample) %>%
    data.frame(row.names = sam$Sample) %>%
    sample_data
st <- st %>%
    select(-Sample) %>%
    data.frame(row.names = st$Sample) %>%
    otu_table(taxa_are_rows = FALSE)
ps <- phyloseq(sam, st)
saveRDS(ps, file.path(data_path, "final", prepend_date("brooks_phyloseq.Rds")))
```

# Add taxonomy and phylogeny using GTDB

```{r tax_setup}
tax_path <- file.path(Sys.getenv("DATA_DIR"), "taxonomy", "gtdb")

gtdb_spec <- cols(
    ssu_gg_blast_bitscore = col_double(),
    ssu_silva_blast_bitscore = col_double()
)
gtdb <- read_tsv(file.path(tax_path, "bac_metadata_r86.tsv"),
    col_types = gtdb_spec,
    na = c("", "NA", "none"))

tree <- ape::read.tree(file.path(tax_path, "bac120_r86.1.tree"))

parse_taxonomy <- function (tax_string) {
    rank_letters <- c("d", "p", "c", "o", "f", "g", "s")
    tax_pattern  <- paste0("(?:", rank_letters, "__([\\w ]+))?") %>%
        paste(collapse = "\\;?")
    tax <- str_match(tax_string, tax_pattern)
    colnames(tax) <- c("Clade", "Kingdom", "Phylum", "Class", "Order", 
        "Family", "Genus", "Species")
    tax %>% as_tibble
}
```

The GTDB metadata table includes Silva, NCBI, and GTDB taxonomies and allows us
to match accessions to leaves in the tree of the closest representative as well
as info about which accessions are themselves in the tree.

Our goal is to link the seven species to their representatives in the tree as
well as get their taxonomy. Which taxonomy do we use? The GTDB taxonomy and to
a lesser extent Silva taxonomy tend to use nonstandard names, so I'll use NCBI
taxonomy for consistency with common usage.

```{r}
gtdb <- gtdb %>%
    mutate(NCBI_species = str_extract(ncbi_taxonomy, "(?<=s__)[\\w ]+"))
species_list <- c("Gardnerella vaginalis", "Atopobium vaginae",
    "Lactobacillus crispatus", "Lactobacillus iners",
    "Prevotella bivia", "Sneathia amnii", "Streptococcus agalactiae")
gtdb0 <- gtdb %>%
    filter(NCBI_species %in% species_list)
# Number of accessions for the seven species:
nrow(gtdb0)
# Check that each species is represented in the tree
gtdb0 %>%
    filter(gtdb_representative == "t") %>%
    select(NCBI_species) %>%
    arrange(NCBI_species) %>%
    distinct
# So let's just restrict to genomes in the phylogeny
gtdb1 <- gtdb0 %>%
    arrange(NCBI_species) %>%
    filter(gtdb_representative == "t")
```

```{r}
gtdb1$ssu_silva_taxonomy %>% head

silva <- gtdb1$ssu_silva_taxonomy %>%
    str_split(";", simplify = TRUE) %>%
    as_tibble
names(silva) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus",
    "Species")
silva <- silva %>%
    mutate(Strain = Species,
        Species = str_extract(Species, "^\\w+ \\w+")) %>%
    add_column(Accession = gtdb1$accession, .before = 1) %>%
    add_column(NCBI_species = gtdb1$NCBI_species, .before = 2)
silva
```

```{r}
gtdb1$ncbi_taxonomy %>% head
ncbi <- gtdb1$ncbi_taxonomy %>%
    parse_taxonomy %>%
    rename(NCBI_taxonomy = Clade)
ncbi <- ncbi %>%
    add_column(Accession = gtdb1$accession, .before = 1) %>%
    add_column(NCBI_species = gtdb1$NCBI_species, .before = 2)
ncbi
```

Which accessions to use for the phylogeny?

Do all share the same taxonomies?
```{r}
# NCBI taxonomy is always the same
gtdb1 %>%
    select(NCBI_species, ncbi_taxonomy) %>%
    distinct

# GTDB taxonomy can vary for three of the species
gtdb1 %>%
    select(NCBI_species, gtdb_taxonomy) %>%
    distinct

# Number of representative genomes to choose from 
gtdb1 %>%
    filter(gtdb_representative == "t") %>%
    group_by(NCBI_species) %>%
    summarize(Num_representatives = n())
```

```{r}
# Silva taxonomy can vary for four of the species, but sometimes this is just
# because of different strain or unidentified species
gtdb1 %>%
    select(NCBI_species, ssu_silva_taxonomy) %>%
    distinct
gtdb1 %>%
    filter(NCBI_species == "Gardnerella vaginalis") %>%
    .$ssu_silva_taxonomy %>%
    unique
```

From http://vmc.vcu.edu/resources/vahmp, we could try to guess what the
correct strians are.  E.g. for Gardnerella:
```{r}
"RS_GCF_000176475.1" %in% gtdb0$accession
```

But it problably doesn't matter which we use. Let's check by first pruning the
tree to the possible representatives and coloring by species. If all strains
with a given NCBI species cluster on the tree then we can just pick a random
representative from each.
```{r}
library(ggtree)

tree0 <- ape::drop.tip(tree, setdiff(tree$tip.label, silva$Accession))

tb <- ncbi %>%
    mutate(taxa = Accession) %>%
    select(taxa, everything())

g <- ggtree(tree0) %<+% tb

g + geom_tiplab(aes(color = NCBI_species)) +
    theme(legend.position = "right")
```

Since all species are monophyletic and all accessions within a species have a
similar length from the root, it doesn't matter which accession we choose for
each species. So I will choose the first (alphabetically by accession) that has
a Silva taxonomy.

```{r}
tb0 <- tb %>%
    group_by(Species) %>%
    top_n(1, Accession)
tree1 <- ape::drop.tip(tree0, setdiff(tree0$tip.label, tb0$Accession))
tree1$tip.label <- tb0$NCBI_species[match(tree1$tip.label, tb0$Accession)]
tb0 <- tb0 %>%
    mutate(taxa = NCBI_species)
(ggtree(tree1) %<+% tb0) + 
    geom_tiplab(aes(color = NCBI_species)) +
    theme(legend.position = "right")
```

Add tree and taxonomy to the phyloseq object:
```{r}
tax <- tb0 %>%
    select(Taxon = taxa, Kingdom:Species)
TAX <- tax %>%
    select(-Taxon) %>%
    as("matrix")
rownames(TAX) <- tax$Taxon
TAX <- tax_table(TAX)
# Replace spaces with underscores:
taxa_names(TAX) <- taxa_names(TAX) %>%
    str_replace(" ", "_")
tree1$tip.label <- tree1$tip.label %>%
    str_replace(" ", "_")
ps0 <- merge_phyloseq(ps, TAX, tree1)
ps0
saveRDS(ps0, file.path(data_path, "final", 
        prepend_date("brooks_phyloseq.Rds")))
```




