---
title: "Extracting and exploring sample metadata"
output:
  html_document:
    toc: true
    toc_float: false
    self_contained: true
---


```{r package_options, include=FALSE, cache=FALSE}
knitr::opts_knit$set(progress=TRUE, verbose=TRUE)
# Global chunk options
knitr::opts_chunk$set(cache=TRUE, autodep=TRUE, echo=TRUE, warning=FALSE,
    message=FALSE, fig.width=10, fig.height=8, include=TRUE)
# R options
options(width=100)
```

# Build a table of sample metadata

The file "SraRunTable.txt" can be found from
https://www.ncbi.nlm.nih.gov/Traces/study/?WebEnv=NCID_1_23816723_130.14.22.76_5555_1524150257_3641866421_0MetA0_S_HStore&query_key=2
by downloading the "RunInfo Table".
The file "biosample_result.txt" can be found from
https://www.ncbi.nlm.nih.gov/bioproject/267701 by following the link to the
list of 240 BioSamples, and going to Send to -> File -> Format: Full (text).
The first file has more info and in a more convenient (tab-separated) form, but
the second file has the useful "Description" entry describing the sample.

```{r}
library(stringr)
library(tidyverse)
library(magrittr)

data_path <- "~/data/vcu_bias"
git_path <- "~/active_research/metagenomics_calibration"
metadata_path <- file.path(git_path, "vcu_bias", "metadata")

## Load the SRA run table with sample info. 
# Despite the .txt extension, this is actually a tab-separated table
sra_tb <- readr::read_tsv(file.path(data_path, "SraRunTable.txt"))
all.equal(tb$Library_Name, tb$Sample_Name)
# Remove non-variable rows and redundant Library_Name (== Sample_Name)
sra_tb <- sra_tb %>%
    select_if(~ (length(unique(.)) > 1)) %>%
    select(-Library_Name)
## Add the sample descriptions as incluced in the BioSample entries
# First, we need to parse the sample identifiers and descriptions from the text
# file.
txt <- readr::read_lines(file.path(data_path, "biosample_result.txt")) 
identifiers <- txt %>% str_subset("Identifier")
idx <- txt %>% str_which("Description:")
descriptions <- txt[idx + 1]
descriptions_alt <- txt %>% str_subset("Sequencing")
all.equal(descriptions, descriptions_alt)
all.equal(length(identifiers), length(descriptions), 240)
# Extract the BioSample identifiers: 
biosamples <- identifiers %>% str_extract("SAMN[0-9]++")
# Match with the description and join with the sra table
desc_tb <- tibble(BioSample = biosamples, Description = descriptions)
tb <- sra_tb %>% left_join(desc_tb, by = "BioSample")
```

The `Sample_Name`s have the format of `TRUTHa_b_c_d` where `a` is the plate,
`b` is the barcode ID, and I'm not yet sure what `c` and `d` are; perhaps these
relate to the experimental design.

Extract info in the Description:
```{r}
# plates <- tb$Description %>% str_match("plate: ([1-6])") %>% .[, 2]
template <- paste0(
    "Sequencing results for a mock community created by mixing",
    " equal amounts of (Cells|DNA|PCR Product) of",
    " \\{(.+) \\}", 
    " with barcode ID: ([0-9]+), barcode: ([A-Z]*), on plate: ([1-6])"
    )
desc_vars <- tb$Description %>% str_match(template) %>%
    as_tibble %>%
    select(-V1, Type = V2, Species = V3, BarcodeID = V4, Barcode = V5, 
        Plate = V6)
all.equal(desc_vars$Species, tb$mixed_culture_contains)
# Calculate the number of species in each mixed culture
desc_vars <- desc_vars %>%
    mutate(number_of_species = str_count(Species, "[A-Z]\\. [a-z]+"))
```

Combine into a new table.
```{r}
tb0 <- bind_cols(tb, desc_vars)
tb1 <- tb0 %>%
    select(Sample = Sample_Name, Type, Contains = Species, 
        Num_species = number_of_species,
        Plate, BarcodeID, Barcode,
        Collection_date = collection_date,
        AvgSpotLen:SRA_Sample,
        Description)
```

```{r}
## Save for future use
readr::write_csv(tb1, file.path(metadata_path, "vcu_sample_metadata.csv"))
```

# Poke around a bit

```{r}
```

