---
title: "Explore the sequence table of Costea et al 2017 Phase III"
output:
  html_document:
    toc: true
    toc_float: false
    code_folding: show
    self_contained: true
---

# Setup

```{r package_options, include=FALSE, cache=FALSE}
# 
knitr::opts_knit$set(progress=TRUE, verbose=TRUE)
# Global chunk options
knitr::opts_chunk$set(cache=TRUE, autodep=TRUE, echo=TRUE, warning=FALSE,
    message=FALSE, fig.width=10, fig.height=8, include=TRUE)
# R options
options(width=100)
# options(digits=2)
```

```{r load_packages, include=T, cache=F}
library(phyloseq)
library(stringr)
library(tidyverse)
library(magrittr)

## Source bias estimation functions
devtools::load_all("~/active_research/src/metacal")

## Avoid confusing behavior from factor conversion
options(stringsAsFactors = FALSE)
```

Load the species-level sequence table.
```{r load_data}
git_path <- file.path("~/active_research/metagenomics_calibration")
data_path <- file.path(git_path, "costea2017", "data")
metadata_path <- file.path(git_path, "costea2017", "metadata")
# "merged_abundance_table_species.tsv" "merged_abundance_table.tsv"        
st <- readr::read_tsv(file.path(data_path,
        "merged_abundance_table_species.tsv"))
samples <- names(st)[-1]
st <- st %>%
    rename(Taxon = ID) %>%
    rename_at(samples, str_extract, "ERR[0-9]+")
taxa <- st$Taxon
st <- st %>% 
    select(-Taxon) %>%
    as("matrix") %>%
    otu_table(taxa_are_rows = TRUE) %>%
    t
taxa_names(st) <- taxa
```
Add the sample data and adjust the sample names from ENA accession to sample
name in Costea et al; then merge into a `phyloseq` object.
```{r load_metadata}
metadata_path <- file.path(git_path, "costea2017", "metadata")
## Sample metadata
# The file sample_description.xlsx is a version of Supplementary Data 3
# reformatted for easier reading into R
phase_dfs <- map(1:3, 
    ~readxl::read_xlsx(file.path(metadata_path, "sample_description_mod.xlsx"), 
        sheet = .))
sam_df <- bind_rows(phase_dfs, .id = "Phase")
## Sequence data info
seq_df <- readr::read_tsv(file.path(metadata_path, "PRJEB14847.tsv"))
## Phase three table
ph3 <- seq_df %>%
    filter(str_detect(library_name, "BYQ"))
sample_names <- ph3$library_name %>%
    str_match("(.*)_DA") %>%
    {.[,2]}
ph3 <- ph3 %>%
    mutate(Sample = sample_names) %>%
    select(Sample, Run_accession = run_accession) %>%
    left_join(sam_df, by="Sample")
# Fix the sample names and add the sample data into a ps object
sample_names(st) <- ph3$Sample[match(sample_names(st), ph3$Run_accession)]
sam <- ph3 %>%
    select(-Sample, -Lab) %>%
    sample_data
sample_names(sam) <- ph3$Sample
ps <- merge_phyloseq(st, sam)
# saveRDS(ps, file.path(data_path, "phase3_phyloseq.Rds"))
ps
sample_sums(ps)
```

# Phylogeny

The tree I'm using is from the `curatedMetagenomicData` package and was
downloaded from
[https://github.com/waldronlab/curatedMetagenomicData/blob/master/inst/extdata/metaphlan2_selected.tree.reroot.nwk.bz2]
<!-- https://groups.google.com/forum/#!msg/metaphlan-users/qEIqLrI37g4/uwFtVeFmAgAJ -->

```{r}
tree <- read_tree(file.path(data_path, "metaphlan2_selected.tree.reroot.nwk"))
# Simplify the species names
tree_species <- tree$tip.label %>% 
    str_match(".*s__(.+)\\|.*") %>% 
    {.[,2]}
# But note there are two tips that didn't have species names
idx <- tree_species %>% is.na %>% which
tree$tip.label[idx]
# So let's prune these
tree0 <- tree
tree0 <- ape::drop.tip(tree0, idx)
tree_species <- tree0$tip.label %>% 
    str_match(".*s__(.+)\\|.*") %>% 
    {.[,2]}
tree0$tip.label <- tree_species

# Some of these species from the ps are in the tree but just with simpler names:
setdiff(taxa_names(ps), tree_species)
taxa <- taxa_names(ps)
taxa %>% str_subset("Streptococcus_mitis")
tree_species %>% str_subset("Streptococcus_mitis")
taxa %>% str_subset("Lactobacillus_casei")
tree_species %>% str_subset("Lactobacillus_casei")
# Other missing taxa are viral and "unclassified"
```

I can manually rename the few mismatched names (or automatically by matching
Genus_species), and when merging, phyloseq will remove the viruses and
"unclassified" taxa. I may want to also do some analyses with the unclassified
taxa such as comparing bias of of the "unclassified" with classified species
within a genus.

```{r}
tree_species0 <- tree_species %>% 
    replace(., . == "Streptococcus_mitis",
        "Streptococcus_mitis_oralis_pneumoniae") %>%
    replace(., . == "Lactobacillus_casei", "Lactobacillus_casei_paracasei")
setdiff(taxa_names(ps), tree_species0)
tree0$tip.label <- tree_species0

ps_tree <- merge_phyloseq(ps, tree0)
# saveRDS(ps_tree, file.path(data_path, "phase3_phyloseq_with_tree.Rds"))
```

# Mock community spike-in

```{r}
ps_mock_only <- subset_samples(ps, Individual == "Mock-only") %>%
    filter_taxa(function (x) sum(x) > 0, prune = TRUE)
# taxa_sums(ps_mock)
# taxa_names(ps_mock) %>% sort
mock_taxa <- c("Blautia_hansenii", "Clostridium_difficile",
    "Clostridium_perfringens", "Clostridium_saccharolyticum",
    "Fusobacterium_nucleatum", "Lactobacillus_plantarum",
    "Prevotella_melaninogenica", "Salmonella_enterica",
    "Vibrio_cholerae", "Yersinia_pseudotuberculosis")
phage <- "Salmonella_phage_Fels_1"
prune_taxa(c(mock_taxa, phage), ps_mock_only) %>% 
    {sample_sums(.) / sample_sums(ps_mock_only)}
prune_taxa(mock_taxa, ps_mock_only) %>% 
    {sample_sums(.) / sample_sums(ps_mock_only)}
```

<!-- Blautia hansenii -->
<!-- Clostridium difficile -->
<!-- Clostridium perfringens -->
<!-- Clostridium saccharolyticum -->
<!-- Fusobacterium nucleatum -->
<!-- Lactobacillus plantarum -->
<!-- Prevotella melaninogenica -->
<!-- Salmonella enterica typhimurium -->
<!-- Vibrio cholerae -->
<!-- Yersinia pseudotuberculosis -->


Let's look at just the mock taxa across all samples.
```{r}
ps0 <- prune_taxa(c(mock_taxa, phage), ps)
```
The fraction of reads from the mock varies greatly across samples, but is
highest in the mock-only samples.
```{r}
(sample_sums(ps0) / sample_sums(ps)) %>% sort
sample_names(ps_mock_only)
```

```{r}
# st <- otu_table(ps_mock) %>% as_tibble
# sam <- sample_data(ps_mock) %>% as_tibble
# tb <- sam %>% left_join(st, by = "Sample")
```

## Get the expected frequency

Skip for now; will have to get clarification from the authors about how to do
this.

## Visualization

There are a few zeros we need to deal with before clr'ing for the PCA. Let's
see what they are:
```{r}
st <- ps0 %>% otu_table
idx <- st %>% {. == 0}
taxa_names(ps0)[apply(idx, 2, any)]
sample_names(ps0)[apply(idx, 1, any)]
sample_data(ps0)[sample_names(ps0)[apply(idx, 1, any)]]
```
I really want the number of mapped reads, not just the frequency, so I can
properly handle zeros. But I'll use the smallest measured frequency as my pseudocount:
```{r}
min_ra <- min(as(st, "matrix")[st>0])
st_clr <- ps0 %>% otu_table %>% 
    transform_sample_counts(function(x) clr(x + min_ra))
pcx <- prcomp(st_clr, scale=F)
p <- plot_pca(ps0, pcx)
ggplot(p$data, aes(x = PC1, y = PC2, color = Protocol, label = Individual)) +
    geom_text() + labs(p$labels[1:2])
```

# Expected abundances of the spike-in

Based on the "ground truth" values of Figure 6. First we have to figure out
what the reported measurements mean.
```{r}
mdf <- readxl::read_xlsx(file.path(metadata_path,
        "composition_of_mock_community.xlsx"), sheet=1)
mdf <- mdf %>% rename(Species = `Bacterial species`)

mdf0 <- mdf %>%
    mutate_at(3:7, function (x) x / gm_mean(x))
mdf1 <- mdf0 %>%
    gather(key = "Type", value = "Ratio", 3:7)
```

```{r}
ggplot(mdf1, aes(x = Species, y = Ratio)) + 
    geom_jitter(width=0.3) +
    facet_wrap(~Type) +
    scale_y_log10() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0),
        axis.title.x = element_blank())
```

Seems like `Bacterial OD in sample` was the target, and the `cells in spike in
Mix` variables are the actual measurment from FACS. The geometrically
normalized and non-normalized values of the two `cells in spike in Mix`
variables are nearly but not quite identical; I'm not sure what the difference
is.
```{r}
mdf[,6:7]
mdf0[,6:7]
```
Note that the species ordering has a few differences from Figure 6 in the
paper, suggesting a possible mislabeling in the Figure or the SI spreadsheet.
```{r}
types <- names(mdf)[6:7]
ggplot(filter(mdf1, Type %in% types), 
    aes(x = fct_reorder(Species, 1/Ratio), y = Ratio)) + 
    geom_jitter(width=0.3) +
    facet_grid(.~Type) +
    scale_y_log10() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0),
        axis.title.x = element_blank())
```
For now, I'll take the average of the clr's of the combined measurements from
these two variables as the "ground truth".
```{r}
types <- names(mdf)[6:7]
truth <- mdf1 %>% filter(Type %in% types) %>%
    mutate(CLR = log2(Ratio)) %>%
    group_by(Species) %>%
    summarize(CLR = mean(CLR))
```
```{r}
ggplot(truth, aes(x = fct_reorder(Species, -CLR), y = 2**CLR)) +
    geom_point() + 
    scale_y_log10() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0),
        axis.title.x = element_blank())
```

Alter species names to match those in my taxomonic profiles:
```{r}
truth0 <- truth %>%
    mutate(Species = str_extract(Species, "\\S+ \\S+"),
        Species = str_replace(Species, " ", "_"))
setequal(c(truth0$Species, "Salmonella_phage_Fels_1"), taxa_names(ps0))
```

I should be able to identify the mislabeling issue by comparing to the observed
data.
```{r}
st <- ps0 %>% otu_table %>% prune_taxa(truth0$Species, .)
min_ra <- min(as(st, "matrix")[st>0])
st_clr <- st %>%
    transform_sample_counts(function(x) clr(x + min_ra))

truth1 <- truth0 %>%
    mutate(Sample = "Truth")

tb <- as_tibble(sample_data(ps0)) %>%
    left_join(as_tibble(st_clr), by = "Sample") %>%
    gather(key = "Species", value = "CLR", truth1$Species)

ggplot(tb, aes(Species, CLR, color = Protocol)) +
    geom_jitter(width=0.3) +
    geom_point(aes(x = Species, y = CLR), data = truth1, color = "black",
        shape = 3, size = 5)
```

My best guess right now is that the species on the x-axis of Figure 6 are
mislabeled.

## Minimal script showing possible mislabeling

```{r}
library(tidyverse)

# "composition_of_mock_community.xlsx" is from the SI data
mdf <- readxl::read_xlsx(file.path(metadata_path,
        "composition_of_mock_community.xlsx"), sheet=1)
mdf <- mdf %>% rename(Species = `Bacterial species`) %>%
    mutate(Species = stringr::str_extract(Species, "\\S+ \\S+"))
mdf0 <- mdf %>%
    select(1, 5:7) %>% 
    mutate_at(-1, function (x) x / gm_mean(x))
mdf1 <- mdf0 %>%
    gather(key = "Type", value = "Ratio", -1)
```

```{r}
fig6_order <- c("Prevotella melaninogenica", "Clostridium perfringens",
    "Salmonella enterica", "Clostridium difficile", "Lactobacillus plantarum",
    "Clostridium saccharolyticum", "Yersinia pseudotuberculosis", 
    "Vibrio cholerae", "Blautia hansenii", "Fusobacterium nucleatum")
setequal(mdf1$Species, fig6_order)
# Specify factor orders for plotting
mdf2 <- mdf1 %>%
    mutate(Species = factor(Species, levels = fig6_order),
        Type = factor(Type, levels = names(mdf)[5:7]))
```

```{r}
ggplot(mdf2, aes(x = Species, y = Ratio)) + 
    geom_jitter(width=0.3) +
    facet_wrap(~Type) +
    scale_y_log10() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0),
        axis.title.x = element_blank())
```

