---
title: "Explore the clade profiles Costea et al 2017 Phase III"
output:
  html_document:
    toc: true
    toc_float: false
    code_folding: show
    self_contained: true
---

# Setup

```{r package_options, include=FALSE, cache=FALSE}
# 
knitr::opts_knit$set(progress=TRUE, verbose=TRUE)
# Global chunk options
knitr::opts_chunk$set(cache=TRUE, autodep=TRUE, echo=TRUE, warning=FALSE,
    message=FALSE, fig.width=10, fig.height=8, include=TRUE)
# R options
options(width=100)
# options(digits=2)
```

```{r load_packages, include=T, cache=F}
library(phyloseq)
library(tidyverse)
library(magrittr)

## Source bias estimation functions
devtools::load_all("~/active_research/src/metacal")

## Avoid confusing behavior from factor conversion
options(stringsAsFactors = FALSE)

brc_data_path <- "~/data/costea2017/brc_out"
git_path <- file.path("~/active_research/metagenomics_calibration")
data_path <- file.path(git_path, "costea2017", "data")
metadata_path <- file.path(git_path, "costea2017", "metadata")

git_path <- file.path("~/active_research/metagenomics_calibration")
data_path <- file.path(git_path, "costea2017", "data")

# Load the sample data and tree
ps <- readRDS(file.path(data_path, "phase3_phyloseq_with_tree.Rds"))
sam <- sample_data(ps)
tree <- phy_tree(ps)
```

```{r}
# fns <- file.path(brc_data_path, list.files(brc_data_path))
# names(fns) <- fns %>%
#     str_extract("ERR[0-9]+")
# profiles <- fns %>%
#     map(parse_clade_profiles, combine_markers = TRUE)
# tb <- profiles %>%
#     bind_rows(.id = "Sample")
# saveRDS(tb, file.path(data_path, "phase3_clade_profiles_tb.Rds"))
```


```{r}
# Load the clade profiles of all samples, and restrict to species profiles
tb <- readRDS(file.path(data_path, "phase3_clade_profiles_tb.Rds")) %>%
    # filter(Rank == "Species") %>%
    # mutate(Taxon =  str_extract(Clade, "(?<=s__)\\w+$"))
    filter(Rank %in% c("Species", "Strain")) %>%
    mutate(Taxon =  str_extract(Clade, "(?<=s__).+$"))
# Fix the sample names
tb$Sample <- sample_names(sam)[match(tb$Sample, sam$Run_accession)]
# Parsed taxonomy
tax <- tb$Clade %>% parse_taxonomy
```

In some species, we only have reads mapped to strain-level
markers. E.g.:
```{r}
tax %>% filter("Blautia_hansenii" == Species)
```

## Check that length always the same for a given clade

Lengths are always the same for a clade across samples:
```{r}
tb %>%
    group_by(Clade) %>%
    summarize(Num = length(unique(Length))) %>%
    filter(Num != 1)
```
I don't yet understand where the quasi-marker disambiguation happens. I could
imagine that if different quasi markers were used in different samples, we
would get different Lengths. 

# Species profiles

## Read counts

Simple method is to sum up reads from species and strains.
```{r}
tb0 <- tb %>%
    left_join(tax, by = "Clade") %>%
    filter(Rank %in% c("Species", "Strain")) %>%
    group_by(Sample, Species) %>%
    summarize(Reads = sum(Reads)) %>%
    ungroup() %>%
    spread(key = Species, value = Reads, fill = 0)
mat <- tb0 %>%
    select(-Sample) %>%
    as("matrix")
rownames(mat) <- tb0$Sample
st <- otu_table(mat, taxa_are_rows = FALSE)
corner(st)
# Make a new ps object
ps.reads <- merge_phyloseq(st, sam, tree, tax)
# saveRDS(ps.reads, file.path(data_path, 
#         prepend_date("phase3_phyloseq_reads.Rds")))
# [1] "2018-05-21_phase3_phyloseq_reads.Rds"
```

## Relative abundance

Get the species relative abundances. To do so, we will use the species-level
weight for length >= 2000 bp and the sum of strain weights if the species is
missing or has a length < 2000

```{r}
tb.tax <- tb %>%
    left_join(tax, by = "Clade")
tb.s <- tb.tax %>%
    filter(Rank == "Species")
# Species abundances from strains:
tb.t <- tb.tax %>%
    filter(Rank == "Strain") %>%
    group_by(Sample, Species) %>%
    summarize(Weight = sum(Weight))
# Species abundances from either tb.s or tb.t
length_threshold <- 2000
tb.st <- full_join(tb.s, tb.t, 
    by = c("Sample", "Species"), suffix = c(".s", ".t")) %>% 
    replace_na(list(Length = 0, Reads = 0, Weight.s = 0, Weight.t = 0))
    # replace_na(list(Length = 0))
tb.st <- tb.st %>%
    mutate(Weight = ifelse(Length >= length_threshold, Weight.s, Weight.t))
# Species relative abundances, compatible with metaphlan after "unclassified" removed
tb.ra <- tb.st %>%
    select(Sample, Clade, Kingdom:Species, Weight) %>%
    group_by(Sample) %>%
    mutate(Weight = Weight / sum(Weight)) %>%
    ungroup()

## Alternative approach, use strain-level estimate if larger
tb.st <- full_join(tb.s, tb.t, 
    by = c("Sample", "Species"), suffix = c(".s", ".t")) %>% 
    replace_na(list(Length = 0, Weight.s = 0, Weight.t = 0))
    # replace_na(list(Length = 0))
tb.st <- tb.st %>%
    mutate(Weight = ifelse((Length < length_threshold) | (Weight.t > Weight.s),
            Weight.t, Weight.s)
        )
# Species relative abundances, compatible with metaphlan after "unclassified" removed
tb.ra <- tb.st %>%
    select(Sample, Clade, Kingdom:Species, Weight) %>%
    group_by(Sample) %>%
    mutate(Weight = Weight / sum(Weight)) %>%
    ungroup()
```

Test on one mock-only sample for simplicity:
```{r}
tb0 <- tb %>%
    filter(Sample == "BYQ_AAAK")
tb.tax <- tb0 %>%
    left_join(tax, by = "Clade")
tb.s <- tb.tax %>%
    filter(Rank == "Species")
# Species abundances from strains:
tb.t <- tb.tax %>%
    filter(Rank == "Strain") %>%
    group_by(Sample, Species) %>%
    summarize(Length = sum(Length), Reads = sum(Reads), Weight = sum(Weight))
# Species abundances from either tb.s or tb.t
length_threshold <- 2000
tb.st <- full_join(tb.s, tb.t, 
    by = c("Sample", "Species"), suffix = c(".s", ".t")) %>% 
    replace_na(list(Length.s = 0, Length.s = 0,
            Reads.s = 0, Reads.t = 0, Weight.s = 0, Weight.t = 0))
    # replace_na(list(Length = 0))
tb.st <- tb.st %>%
    mutate(Weight = ifelse(Length.s >= length_threshold, Weight.s, Weight.t))
# Species relative abundances, compatible with metaphlan after "unclassified" removed
tb.ra <- tb.st %>%
    select(Sample, Clade, Kingdom:Species, Weight) %>%
    group_by(Sample) %>%
    mutate(Weight = Weight / sum(Weight)) %>%
    ungroup()
##### Explore:
tb.st0 <- select(tb.st, Species, 
    Length.s:Weight.s,
    Length.t:Weight)
tb.st0 %>% 
    filter(Weight.t > Weight.s) %>% 
    arrange(desc(Weight))
tb.st0 %>% 
    filter(Length.s > length_threshold, Weight.t > Weight.s) %>% 
    arrange(desc(Weight))
```
There are many cases where the weight from strains is greater than the weight
from species, despite the large total length of species-level markers, but the
differences usually seem to be fairly small. In such cases, metaphlan2 uses the
strain Weight.

# Mock community

Goal is to check that we get the same results as when we used the stock metaphlan output.

For some species, there may be reads mapped at the strain level but not the
species level, and so to calculate species abundances we'll need to account for
this.
```{r}

```

Let's get a phyloseq object that is compatible with the one I used in mock_taxa.Rmd.
```{r, eval = F}
## Make a species table
st.ra <- tb.ra %>%
    select(Sample, Taxon = Species, Weight) %>%
    spread(key = Taxon, value = Weight, fill = 0)
# Get a ps object
mat <- st.ra %>%
    select(-Sample) %>%
    as("matrix")
rownames(mat) <- st.ra$Sample
st.ps <- otu_table(mat, taxa_are_rows= F)
sam <- ph3 %>%
    select(-Sample, -run_accession) %>%
    sample_data
sample_names(sam) <- ph3$Sample
ps <- merge_phyloseq(st.ps, sam)
```

```{r}
mock_taxa <- c("Blautia_hansenii", "Clostridium_difficile",
    "Clostridium_perfringens", "Clostridium_saccharolyticum",
    "Fusobacterium_nucleatum", "Lactobacillus_plantarum",
    "Prevotella_melaninogenica", "Salmonella_enterica",
    "Vibrio_cholerae", "Yersinia_pseudotuberculosis")
all(mock_taxa %in% taxa_names(ps))
```

