---
title: "Pool bias estimates across individuals"
output:
  html_document:
    toc: true
    toc_float: false
    code_folding: hide
    self_contained: true
---

# Setup

```{r package_options, include=FALSE, cache=FALSE}
# 
knitr::opts_knit$set(progress=TRUE, verbose=TRUE)
# Global chunk options
knitr::opts_chunk$set(cache=TRUE, autodep=TRUE, echo=TRUE, warning=FALSE,
    message=FALSE, fig.width=10, fig.height=8, include=TRUE)
# R options
options(width=100)
# options(digits=2)
```

```{r load_packages, include=T, cache=F}
library(phyloseq)
library(tidyverse)
library(magrittr)
# library(ggtree)
library(ggthemes)

## Source bias estimation functions
devtools::load_all("~/active_research/src/metacal")

## Avoid confusing behavior from factor conversion
options(stringsAsFactors = FALSE)
```

```{r load_data}
git_path <- file.path("~/active_research/metagenomics_calibration")
data_path <- file.path(git_path, "costea2017", "data")
ps <- readRDS(file.path(data_path, "2018-05-22_phase3_phyloseq_reads.Rds"))
ps.ra <- readRDS(file.path(data_path, "phase3_phyloseq_with_tree.Rds"))
# Get rid of the samples of Individual = A or B b/c these were only extracted
# by one protocol
ps <- subset_samples(ps, Individual %in% as.character(1:8))
ps.ra <- subset_samples(ps.ra, Individual %in% as.character(1:8))
# Species in the mock community
mock_taxa <- c("Blautia_hansenii", "Clostridium_difficile",
    "Clostridium_perfringens", "Clostridium_saccharolyticum",
    "Fusobacterium_nucleatum", "Lactobacillus_plantarum",
    "Prevotella_melaninogenica", "Salmonella_enterica",
    "Vibrio_cholerae", "Yersinia_pseudotuberculosis")
# denom for clr
denom <- setdiff(mock_taxa, "Blautia_hansenii")
```

# Simple method: median of clrs (no sampling)

## Mock community

First, let's set with just the mock:
```{r}
ps.mock <- prune_taxa(mock_taxa, ps)
# Get observed count tables.
taxa <- taxa_names(ps.mock)
tb <- ps.mock %>% as_tibble %>%
    select(Protocol, Individual, taxa) %>%
    arrange(Protocol, Individual)
# corner(tb)
observed.tar <- tb %>% 
    filter(Protocol == "H") %>%
    select(taxa) %>%
    as("matrix")
observed.ref <- tb %>% 
    filter(Protocol == "W") %>%
    select(taxa) %>%
    as("matrix")
```

Biases by simplest possible method:
```{r}
observed.tar.clr <- observed.tar %>%
    apply(1, clr) %>%
    t
observed.ref.clr <- observed.ref %>%
    apply(1, clr) %>%
    t
bias <- (observed.tar.clr - observed.ref.clr) %>%
    as_tibble %>%
    summarize_all(median) %>%
    gather(key = "Taxon", value = "Log2_bias")
bias %>%
    mutate(Ratio = 2^Log2_bias,
        Log10_bias = log10(Ratio))
```
These look to be correct.

## More taxa

First, filter to taxa that show up in a minimum (relative) number of reads,
below which is likely contamination, and a minimum relative abundance.
```{r}
ps.r <- transform_sample_counts(ps, function (x) x / sum(x))
filter_taxa(ps.r, function(x) max(x) > 5e-4) %>% sum
ps0 <- ( filter_taxa(ps.r, function(x) max(x) > 5e-4) & 
    filter_taxa(ps.ra, function(x) max(x) > 5e-4) ) %>%
    prune_taxa(ps)
tax <- tax_table(ps0) %>% as_tibble
filter(tax, Class == "Gammaproteobacteria")
```
But, assume that for a taxon to be present in an individual, it must show up in
at least a fraction 5e-4 of all reads.
```{r}
taxa <- taxa_names(ps0)
tb <- ps.r %>% as_tibble %>% select(Individual, taxa)
tb0 <- tb %>%
    group_by(Individual) %>%
    summarize_at(taxa, ~max(.) > 5e-4)
corner(tb0)
# Matrix for screening out taxa from individuals
screen_mat <- as(select(tb0, taxa), "matrix")
```
Next, get observed count tables.
```{r}
tb <- ps0 %>% as_tibble %>%
    select(Protocol, Individual, taxa) %>%
    arrange(Protocol, Individual)
# corner(tb)
observed.tar <- tb %>% 
    filter(Protocol == "H") %>%
    select(taxa) %>%
    as("matrix")
observed.ref <- tb %>% 
    filter(Protocol == "W") %>%
    select(taxa) %>%
    as("matrix")
# corner(observed.tar, 5, 4)
# corner(observed.ref, 5, 4)
## Add a pseudocount to deal with zeros
observed.tar <- observed.tar + 1
observed.ref <- observed.ref + 1
## Screen out taxa that are likely to be contaminants
observed.tar <- observed.tar * screen_mat
observed.ref <- observed.ref * screen_mat
# corner(observed.tar, 5, 4)
# corner(observed.ref, 5, 4)
```

Biases by simplest possible method:
```{r}
observed.tar.clr <- observed.tar %>%
    apply(1, clr, denom = mock_taxa) %>%
    t
observed.ref.clr <- observed.ref %>%
    apply(1, clr, denom = mock_taxa) %>%
    t
# observed.tar.clr[1:5, mock_taxa]
# observed.ref.clr[1:5, mock_taxa]
bias <- (observed.tar.clr - observed.ref.clr) %>%
    as_tibble %>%
    summarize_all(~ median(.[is.finite(.)])) %>%
    gather(key = "Taxon", value = "Log2_bias") %>%
    mutate(Ratio = 2^Log2_bias,
        Log10_bias = log10(Ratio))
bias %>%
    filter(Taxon %in% mock_taxa)
```

